### DISCUSSION TITLE
Inter-VM communication using virtual network devices?

### DISCUSSION THREAD

**Discussion by yanjiew1** [bao-project/bao-hypervisor#104] [General]
Hello, I am a CS graduate student at National Taiwan University.

Currently, it seems that the only way for inter-VM communication is to use shared memory and inter-VM interrupts.
I would like to ask whether there are any plans to implement virtual network devices and virtual switch in bao-hypervisor.

**Comment by josecm** [comment]
Hello @yanjiew1!

The shared memory + inter-VM interrupts are only the base mechanisms to build more complex communication on top, at the guest level, not inside bao. Including virtual networks.

Currently, we are working on virtio support with backends in separate virtual machines, which will likely facilitate the construction of virtual networks and virtual switches on a service VM. 

Nevertheless, if your requirements are just single 1-to-1 communication over a virtual network we have already implemented such a proof-of-concept in the past based on Jailhouse's IVSHMEM-net linux drivers. You can take a look at it in [this demo](https://github.com/bao-project/bao-demos/tree/pseudo-ivshmem-net) (although this is using a significantly older version of Bao).

Let me know if this was any help. Also, if you are available to collaborate on this topic, we could share with you our current virtio support, so you could play around with virtio-net support.

**Reply by yanjiew1** [reply]
I see. Thank you!
